File: apps/agent/src/agent/tools/expand-via-graph.ts
import { entitiesCol, relationsCol, entityKey } from "../../lib/graph.js";
import { getCollections } from "../../lib/mongo.js";
import { ObjectId } from "mongodb";
import type { Citation, ToolDef } from "../types.js";

/**
 * Graph-augmented retrieval.
 *
 * Given an entity name (or a list of them), walk the memory graph:
 *   1. Resolve each name → entity record (people / projects / topics).
 *   2. For depth ≥ 1, collect neighbours via the `relations` collection.
 *   3. Pull every chunkId referenced by the entity-set.
 *   4. Hydrate to full chunks and return as citations.
 *
 * This complements `search_memory` — that finds chunks that *look like* your
 * query; this finds chunks that are *about the same people* even when the
 * wording is different. Combine them for true graph-RAG.
 */
export const expandViaGraphTool: ToolDef = {
  declaration: {
    name: "expand_via_graph",
    description:
      "Walk the memory graph to fetch chunks connected to a specific person, project, or topic — even chunks that don't share keywords with the original query. Use this AFTER search_memory when you've found a key entity and want to pull in everything else they're connected to. Returns chunks plus the path of entities traversed.",
    parameters: {
      type: "OBJECT",
      properties: {
        entity: {
          type: "string",
          description: "Canonical name of the seed entity (e.g. 'Sarah Okafor', 'Lantern', 'Q3 planning'). Case-insensitive.",
        },
        entities: {
          type: "array",
          items: { type: "string" },
          description: "Optional: multiple seed entities to expand from in parallel.",
        },
        depth: {
          type: "integer",
          description: "Hops to traverse along relations. 1 = seed + immediate neighbours. 2 = also neighbours-of-neighbours. Default 1, max 2.",
        },
        kinds: {
          type: "array",
          items: { type: "string", enum: ["owes", "works_with", "manages", "discusses"] },
          description: "Optional: restrict traversal to specific relation kinds.",
        },
        limit: {
          type: "integer",
          description: "Max chunks to return. Default 12, max 30.",
        },
      },
    },
  },
  handler: async (args) => {
    const seeds: string[] = [];
    if (typeof args["entity"] === "string" && args["entity"].trim()) {
      seeds.push(args["entity"].trim());
    }
    if (Array.isArray(args["entities"])) {
      for (const e of args["entities"]) {
        if (typeof e === "string" && e.trim()) seeds.push(e.trim());
      }
    }
    if (seeds.length === 0) {
      return { ok: false, error: "at least one entity name is required" };
    }
    const depth = clampInt(args["depth"], 1, 1, 2);
    const limit = clampInt(args["limit"], 12, 1, 30);
    const kinds: Array<"owes" | "works_with" | "manages" | "discusses"> | null =
      Array.isArray(args["kinds"]) && args["kinds"].length > 0
        ? args["kinds"].filter(
            (k): k is "owes" | "works_with" | "manages" | "discusses" =>
              k === "owes" || k === "works_with" || k === "manages" || k === "discusses",
          )
        : null;

    try {
      const ents = await entitiesCol();
      const rels = await relationsCol();

      // Resolve seed names → entity records (case-insensitive contains match).
      const seedKeys = new Set<string>();
      const seedEntities: Array<{ key: string; name: string; kind: string }> = [];
      for (const s of seeds) {
        const direct = entityKey(s);
        // Try exact key match first, then partial name match.
        const exact = await ents.findOne({ key: direct });
        if (exact) {
          seedKeys.add(exact.key);
          seedEntities.push({ key: exact.key, name: exact.name, kind: exact.kind });
          continue;
        }
        const fuzzy = await ents.findOne({ name: { $regex: s, $options: "i" } });
        if (fuzzy) {
          seedKeys.add(fuzzy.key);
          seedEntities.push({ key: fuzzy.key, name: fuzzy.name, kind: fuzzy.kind });
        }
      }

      if (seedEntities.length === 0) {
        return {
          ok: true,
          data: { resolved: [], chunks: [], traversed: [], chunksFound: 0 },
          summary: `no entities matched: ${seeds.join(", ")}`,
        };
      }

      // BFS traversal across the relations graph.
      const visited = new Set<string>(seedKeys);
      const traversedRels: Array<{ from: string; to: string; kind: string }> = [];
      let frontier = Array.from(seedKeys);

      for (let d = 0; d < depth; d++) {
        if (frontier.length === 0) break;
        const relFilter: Record<string, unknown> = {
          $or: [{ from: { $in: frontier } }, { to: { $in: frontier } }],
        };
        if (kinds) relFilter["kind"] = { $in: kinds };
        const found = await rels.find(relFilter, { limit: 200 }).toArray();

        const nextFrontier: string[] = [];
        for (const r of found) {
          const other = frontier.includes(r.from) ? r.to : r.from;
          if (!visited.has(other)) {
            visited.add(other);
            nextFrontier.push(other);
          }
          traversedRels.push({ from: r.from, to: r.to, kind: r.kind });
        }
        frontier = nextFrontier;
      }

      // Pull chunkIds referenced by ALL visited entities.
      const allEntities = await ents.find({ key: { $in: Array.from(visited) } }).toArray();
      const chunkIdSet = new Set<string>();
      for (const e of allEntities) {
        for (const cid of e.chunkIds ?? []) chunkIdSet.add(cid);
      }

      // Hydrate to full chunks (with metadata for citations).
      const { chunks: chunksCol } = await getCollections();
      const chunkOids: ObjectId[] = [];
      for (const cid of chunkIdSet) {
        if (ObjectId.isValid(cid)) chunkOids.push(new ObjectId(cid));
      }
      const chunks = chunkOids.length > 0
        ? await chunksCol
            .find(
              { _id: { $in: chunkOids } },
              {
                projection: {
                  _id: 1,
                  documentId: 1,
                  source: 1,
                  title: 1,
                  text: 1,
                  ordinal: 1,
                  metadata: 1,
                },
                limit,
              },
            )
            .toArray()
        : [];

      const citations: Citation[] = chunks.map((c) => ({
        chunkId: String(c._id),
        documentId: String(c.documentId),
        source: c.source as Citation["source"],
        title: String(c.title),
        score: 1, // graph-fetched, no relevance score
        ordinal: Number(c.ordinal ?? 0),
        text: typeof c.text === "string" ? c.text.slice(0, 300) : undefined,
      }));

      const data = {
        resolved: seedEntities,
        traversed: allEntities.map((e) => ({ name: e.name, kind: e.kind, key: e.key })),
        relations: traversedRels.slice(0, 50),
        chunksFound: chunkIdSet.size,
        chunks: chunks.map((c) => ({
          chunkId: String(c._id),
          title: c.title,
          source: c.source,
          ordinal: c.ordinal,
          text: c.text,
          metadata: c.metadata,
        })),
      };

      return {
        ok: true,
        data,
        citations,
        summary: `graph expand · ${seedEntities.length} seed → ${allEntities.length} entities · depth ${depth} · ${chunks.length}/${chunkIdSet.size} chunks`,
      };
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      return { ok: false, error: msg };
    }
  },
};

function clampInt(raw: unknown, dflt: number, lo: number, hi: number): number {
  const n = typeof raw === "number" ? raw : Number(raw);
  if (!Number.isFinite(n)) return dflt;
  return Math.max(lo, Math.min(hi, Math.floor(n)));
}

File: apps/agent-py/app/agent/tools/expand_via_graph.py
"""expand_via_graph — port of apps/agent/src/agent/tools/expand-via-graph.ts."""
from __future__ import annotations

from bson import ObjectId

from app.agent.types import ToolDef
from app.db.mongo import chunks
from app.lib.graph import entities_col, entity_key, relations_col

_KINDS = ("owes", "works_with", "manages", "discusses")

_DECL = {
    "name": "expand_via_graph",
    "description": (
        "Walk the memory graph to fetch chunks connected to a specific person, project, or topic — "
        "even chunks that don't share keywords with the original query. Use this AFTER search_memory "
        "when you've found a key entity and want to pull in everything else they're connected to. "
        "Returns chunks plus the path of entities traversed."
    ),
    "parameters": {
        "type": "OBJECT",
        "properties": {
            "entity": {"type": "string", "description": "Canonical name of the seed entity (case-insensitive)."},
            "entities": {"type": "array", "items": {"type": "string"}, "description": "Optional: multiple seed entities."},
            "depth": {"type": "integer", "description": "Hops to traverse. 1 = seed + neighbours. Default 1, max 2."},
            "kinds": {"type": "array", "items": {"type": "string", "enum": list(_KINDS)},
                      "description": "Optional: restrict traversal to specific relation kinds."},
            "limit": {"type": "integer", "description": "Max chunks to return. Default 12, max 30."},
        },
    },
}


def _clamp(raw, default, lo, hi):
    try:
        return max(lo, min(hi, int(raw)))
    except (TypeError, ValueError):
        return default


async def _handler(args: dict, ctx: dict | None = None) -> dict:
    seeds: list[str] = []
    if isinstance(args.get("entity"), str) and args["entity"].strip():
        seeds.append(args["entity"].strip())
    if isinstance(args.get("entities"), list):
        seeds += [e.strip() for e in args["entities"] if isinstance(e, str) and e.strip()]
    if not seeds:
        return {"ok": False, "error": "at least one entity name is required"}
    depth = _clamp(args.get("depth"), 1, 1, 2)
    limit = _clamp(args.get("limit"), 12, 1, 30)
    kinds = [k for k in (args.get("kinds") or []) if k in _KINDS] or None

    try:
        ents, rels = entities_col(), relations_col()
        seed_keys: set[str] = set()
        seed_entities: list[dict] = []
        for s in seeds:
            exact = await ents.find_one({"key": entity_key(s)})
            if exact:
                seed_keys.add(exact["key"])
                seed_entities.append({"key": exact["key"], "name": exact["name"], "kind": exact["kind"]})
                continue
            fuzzy = await ents.find_one({"name": {"$regex": s, "$options": "i"}})
            if fuzzy:
                seed_keys.add(fuzzy["key"])
                seed_entities.append({"key": fuzzy["key"], "name": fuzzy["name"], "kind": fuzzy["kind"]})

        if not seed_entities:
            return {"ok": True, "data": {"resolved": [], "chunks": [], "traversed": [], "chunksFound": 0},
                    "summary": f"no entities matched: {', '.join(seeds)}"}

        visited = set(seed_keys)
        traversed_rels: list[dict] = []
        frontier = list(seed_keys)
        for _ in range(depth):
            if not frontier:
                break
            rel_flt: dict = {"$or": [{"from": {"$in": frontier}}, {"to": {"$in": frontier}}]}
            if kinds:
                rel_flt["kind"] = {"$in": kinds}
            found = await rels.find(rel_flt, limit=200).to_list(length=None)
            next_frontier: list[str] = []
            for r in found:
                other = r["to"] if r["from"] in frontier else r["from"]
                if other not in visited:
                    visited.add(other)
                    next_frontier.append(other)
                traversed_rels.append({"from": r["from"], "to": r["to"], "kind": r["kind"]})
            frontier = next_frontier

        all_entities = await ents.find({"key": {"$in": list(visited)}}).to_list(length=None)
        chunk_ids: set[str] = set()
        for e in all_entities:
            for cid in e.get("chunkIds", []):
                chunk_ids.add(cid)

        oids = [ObjectId(cid) for cid in chunk_ids if ObjectId.is_valid(cid)]
        rows = await chunks().find(
            {"_id": {"$in": oids}},
            projection={"_id": 1, "documentId": 1, "source": 1, "title": 1, "text": 1, "ordinal": 1, "metadata": 1},
            limit=limit,
        ).to_list(length=None) if oids else []

        citations = [{
            "chunkId": str(c["_id"]), "documentId": str(c.get("documentId", "")), "source": c.get("source"),
            "title": str(c.get("title", "")), "score": 1, "ordinal": int(c.get("ordinal", 0)),
            "text": (c.get("text") or "")[:300],
        } for c in rows]

        data = {
            "resolved": seed_entities,
            "traversed": [{"name": e["name"], "kind": e["kind"], "key": e["key"]} for e in all_entities],
            "relations": traversed_rels[:50],
            "chunksFound": len(chunk_ids),
            "chunks": [{
                "chunkId": str(c["_id"]), "title": c.get("title"), "source": c.get("source"),
                "ordinal": c.get("ordinal"), "text": c.get("text"), "metadata": c.get("metadata"),
            } for c in rows],
        }
        return {"ok": True, "data": data, "citations": citations,
                "summary": f"graph expand · {len(seed_entities)} seed → {len(all_entities)} entities · depth {depth} · {len(rows)}/{len(chunk_ids)} chunks"}
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "error": str(err)}


tool = ToolDef(declaration=_DECL, handler=_handler)
